lib/block_view_utils.go

```diff
@@ -40,6 +43,25 @@ func (safeUtxoView *SafeUtxoView) ConnectTransaction(
 	_fees uint64,
 	_err error,
 ) {
+	var revertBackupToPrimary bool
+	// If a transaction panics when connecting to the primary view, then
+	// we know it is invalid and we should revert the primary view to the
+	// backup view. If the transaction panics when connecting to the backup
+	// view, then we should revert the backup view to the primary view and
+	// return a valid txn response.
+	defer func() {
+		if r := recover(); r != nil {
+			if revertBackupToPrimary {
+				safeUtxoView.backupView = safeUtxoView.primaryView.CopyUtxoView()
+				return
+			}
+			glog.Errorf("safeUtxoView.ConnectTransaction: Recovered from panic: %v", r)
+			_utxoOps = nil
+			_totalInput, _totalOutput, _fees = 0, 0, 0
+			_err = errors.Errorf("ConnectTransaction: Recovered from panic: %v", r)
+			safeUtxoView.primaryView = safeUtxoView.backupView.CopyUtxoView()
```

Comment:
Hmm... is this the only way to adjust the return values? Normally we always try to return explicitly; we don't rely on setting return values to the return arguments because it's error-prone.
```
return nil, 0, 0, 0, errors.Errorf("ConnectTransaction: Recovered from panic: %v", r)
```